File: src/Solutions/Euler054.hs
module Solutions.Euler054 where
import Utils.Input(getInput)
import Data.Char
import Data.List
import Data.Ord
import Data.Maybe

type Value = Int -- 11 = J, 12 = Q, 13 = K, 14 = A
jack = 11
queen = 12
king = 13
ace = 14
data Suit = Diamond | Club | Heart | Spade deriving (Eq, Show)
data Card = Card {value :: Value, suit :: Suit} deriving (Eq, Show)
instance Ord Card where compare c1 c2 = compare (value c1) (value c2)

parseCard :: String -> Card
parseCard [v, s] = Card v' s' where
    v' = case v of
        'T' -> 10
        'J' -> 11
        'Q' -> 12
        'K' -> 13
        'A' -> 14
        _ -> digitToInt v
    s' = case s of
        'D' -> Diamond
        'C' -> Club
        'H' -> Heart
        'S' -> Spade

parseHand :: String -> [Card]
parseHand handStr = [c1, c2, c3, c4, c5] where
    c1 = parseCard $ take 2 handStr
    c2 = parseCard $ (take 2 . drop 3) handStr
    c3 = parseCard $ (take 2 . drop 6) handStr
    c4 = parseCard $ (take 2 . drop 9) handStr
    c5 = parseCard $ (take 2 . drop 12) handStr

parseTwoHands :: String -> ([Card], [Card])
parseTwoHands handsStr = (hand1, hand2) where
    hand1 = parseHand handsStr
    hand2 = parseHand $ drop 15 handsStr

data HandRank
    = HighCard -- remaining: all 5 cards sorted
    | OnePair Value -- remaining: 3 cards sorted
    | TwoPairs Value Value -- v1 > v2, remaining: last card
    | Three Value -- remaining: 2 cards sorted
    | Straight Value -- highest value
    | Flush -- remaining: all 5 cards
    | FullHouse Value Value -- 3 + 2
    | Four Value -- remaining: last card
    | StraightFlush Value -- highest value
    | RoyalFlush
    deriving (Eq, Ord, Show)

data Hand = Hand { handRank :: HandRank, remaining :: [Value] }
    deriving (Eq, Ord, Show)

straight :: [Card] -> Bool
straight [card1, card2, card3, card4, card5] =
    let h = value card1
    in value card2 == h - 1 && value card3 == h - 2 && value card4 == h - 3 && value card5 == h - 4

flush :: [Card] -> Bool
flush hand = let s = suit (head hand) in all (\c -> suit c == s) hand

four :: [Card] -> Maybe Hand
four [card1, card2, card3, card4, card5]
    | value card2 /= value card3 || value card3 /= value card4 = Nothing
    | value card1 == value card3 = Just $ Hand (Four (value card3)) [value card5]
    | value card5 == value card3 = Just $ Hand (Four (value card3)) [value card1]
    | otherwise = Nothing

pos3 :: [Card] -> Maybe Int
pos3 [card1, card2, card3, card4, card5] -- 3rd card must be in three of a kind
    | p1 && p2 = Just 1
    | p2 && p4 = Just 2
    | p4 && p5 = Just 3
    | otherwise = Nothing
    where p1 = value card1 == value card3
          p2 = value card2 == value card3
          p4 = value card4 == value card3
          p5 = value card5 == value card3

evalHand :: [Card] -> Hand
evalHand unsortedHand =
    let hand@[card1, card2, card3, card4, card5] = sortOn Data.Ord.Down unsortedHand :: [Card] in -- sorted from highest to lowest
    case (straight hand, flush hand) of
        (True, True) -> if value card1 == ace
            then Hand RoyalFlush []
            else Hand (StraightFlush $ value card1) []
        (True, False) -> Hand (Straight $ value card1) []
        (False, True) -> Hand Flush $ map value hand
        (False, False) -> flip fromMaybe (four hand) $
            case pos3 hand of
                Just 2 -> Hand (Three $ value card2) (map value [card1, card5])
                Just 1 -> if value card4 == value card5
                    then Hand (FullHouse (value card1) (value card4)) []
                    else Hand (Three $ value card1) (map value [card4, card5])
                Just 3 -> if value card1 == value card2
                    then Hand (FullHouse (value card3) (value card1)) []
                    else Hand (Three $ value card3) (map value [card1, card2])
                Nothing ->
                    let p1 = value card1 == value card2
                        p2 = value card2 == value card3
                        p3 = value card3 == value card4
                        p4 = value card4 == value card5 in
                        case (p1, p2, p3, p4) of
                            (True, False, True, False) -> Hand (TwoPairs (value card1) (value card3)) [value card5]
                            (False, True, False, True) -> Hand (TwoPairs (value card2) (value card4)) [value card1]
                            (True, False, False, True) -> Hand (TwoPairs (value card1) (value card4)) [value card3]
                            (True, False, False, False) -> Hand (OnePair (value card1)) (map value [card3, card4, card5])
                            (False, True, False, False) -> Hand (OnePair (value card2)) (map value [card1, card4, card5])
                            (False, False, True, False) -> Hand (OnePair (value card3)) (map value [card1, card2, card5])
                            (False, False, False, True) -> Hand (OnePair (value card4)) (map value [card1, card2, card3])
                            _ -> Hand HighCard (map value hand)

evalLine :: String -> (Hand, Hand)
evalLine = (\(a, b) -> (evalHand a, evalHand b)) . parseTwoHands

euler054 :: IO String
euler054 = show . length . filter (uncurry (>)) . map evalLine . lines <$> getInput 54

File: rust/src/solutions/euler054.rs
use itertools::Itertools;
use std::cmp::Ordering;

type Value = usize;
const ACE: Value = 14;
const KING: Value = 13;
const QUEEN: Value = 12;
const JACK: Value = 11;

fn parse_value(c: char) -> Value {
    match c {
        'A' => ACE,
        'K' => KING,
        'Q' => QUEEN,
        'J' => JACK,
        'T' => 10,
        _ => c.to_digit(10).unwrap() as usize,
    }
}

#[derive(Eq, PartialEq)]
enum Suit {
    Diamond,
    Club,
    Heart,
    Spade,
}

fn parse_suit(c: char) -> Suit {
    match c {
        'D' => Suit::Diamond,
        'C' => Suit::Club,
        'H' => Suit::Heart,
        _ => Suit::Spade,
    }
}

#[derive(Eq, PartialEq)]
struct Card {
    pub value: Value,
    pub suit: Suit,
}

impl Ord for Card {
    fn cmp(&self, other: &Self) -> Ordering {
        self.value.cmp(&other.value)
    }
}

impl PartialOrd for Card {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn parse_card(s: &[char]) -> Card {
    Card {
        value: parse_value(s[0]),
        suit: parse_suit(s[1]),
    }
}

struct Hand([Card; 5]);

fn parse_hand(s: &[char]) -> Hand {
    let mut cards = [
        parse_card(&s[0..2]),
        parse_card(&s[3..5]),
        parse_card(&s[6..8]),
        parse_card(&s[9..11]),
        parse_card(&s[12..14]),
    ];
    cards.sort();
    cards.reverse();
    Hand(cards)
}

fn parse_hands(s: &[char]) -> (Hand, Hand) {
    (parse_hand(&s[..15]), parse_hand(&s[15..]))
}

// Card fields represent remaining (not used in) cards
// Those cards are sorted with respect to their value

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct HighCard {
    pub remaining: [Value; 5],
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct OnePair {
    pub pair: Value,
    pub remaining: [Value; 3],
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct TwoPairs {
    pub higher_pair: Value,
    pub lower_pair: Value,
    pub remaining: Value,
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct Three {
    pub three: Value,
    pub remaining: [Value; 2],
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct Straight {
    pub highest: Value,
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct Flush {
    pub remaining: [Value; 5],
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct FullHouse {
    pub three: Value,
    pub two: Value,
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct Four {
    pub four: Value,
    pub remaining: Value,
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
struct StraightFlush {
    pub highest: Value,
}

#[derive(Ord, PartialOrd, Eq, PartialEq)]
enum HandRank {
    HighCard(HighCard),
    OnePair(OnePair),
    TwoPairs(TwoPairs), // value1 > value2
    Three(Three),
    Straight(Straight), // highest value
    Flush(Flush),
    FullHouse(FullHouse), // value1 x 3 + value2 x 2
    Four(Four),
    StraightFlush(StraightFlush), // highest value
    RoyalFlush,
}

fn find_straight(Hand(cards): &Hand) -> Option<Straight> {
    let values: Vec<Value> = cards.iter().map(|c| c.value).collect();
    if (1..5).all(|i| values[i] == values[0] - i) {
        Some(Straight { highest: values[0] })
    } else {
        None
    }
}

fn find_flush(Hand(cards): &Hand) -> Option<Flush> {
    if cards.iter().map(|c| c.value).all_equal() {
        Some(Flush {
            remaining: [
                cards[0].value,
                cards[1].value,
                cards[2].value,
                cards[3].value,
                cards[4].value,
            ],
        })
    } else {
        None
    }
}

fn find_four(Hand(cards): &Hand) -> Option<Four> {
    if cards[0..4].iter().map(|c| c.value).all_equal() {
        Some(Four {
            four: cards[2].value,
            remaining: cards[4].value,
        })
    } else if cards[1..5].iter().map(|c| c.value).all_equal() {
        Some(Four {
            four: cards[2].value,
            remaining: cards[0].value,
        })
    } else {
        None
    }
}

fn rank_low_hand(Hand([card1, card2, card3, card4, card5]): &Hand) -> HandRank {
    let v = card3.value;
    if [v, card1.value, card2.value].iter().all_equal() {
        if card4.value == card5.value {
            HandRank::FullHouse(FullHouse {
                three: v,
                two: card4.value,
            })
        } else {
            HandRank::Three(Three {
                three: v,
                remaining: [card4.value, card5.value],
            })
        }
    } else if [v, card4.value, card2.value].iter().all_equal() {
        HandRank::Three(Three {
            three: v,
            remaining: [card1.value, card5.value],
        })
    } else if [v, card4.value, card5.value].iter().all_equal() {
        if card1.value == card2.value {
            HandRank::FullHouse(FullHouse {
                three: v,
                two: card1.value,
            })
        } else {
            HandRank::Three(Three {
                three: v,
                remaining: [card1.value, card2.value],
            })
        }
    } else {
        let p1 = card1.value == card2.value;
        let p2 = card2.value == card3.value;
        let p3 = card3.value == card4.value;
        let p4 = card4.value == card5.value;
        match (p1, p2, p3, p4) {
            (true, false, true, false) => HandRank::TwoPairs(TwoPairs {
                higher_pair: card1.value,
                lower_pair: card3.value,
                remaining: card5.value,
            }),
            (false, true, false, true) => HandRank::TwoPairs(TwoPairs {
                higher_pair: card2.value,
                lower_pair: card4.value,
                remaining: card1.value,
            }),
            (true, false, false, true) => HandRank::TwoPairs(TwoPairs {
                higher_pair: card1.value,
                lower_pair: card4.value,
                remaining: card3.value,
            }),
            (true, false, false, false) => HandRank::OnePair(OnePair {
                pair: card1.value,
                remaining: [card3.value, card4.value, card5.value],
            }),
            (false, true, false, false) => HandRank::OnePair(OnePair {
                pair: card2.value,
                remaining: [card1.value, card4.value, card5.value],
            }),
            (false, false, true, false) => HandRank::OnePair(OnePair {
                pair: card3.value,
                remaining: [card1.value, card2.value, card5.value],
            }),
            (false, false, false, true) => HandRank::OnePair(OnePair {
                pair: card4.value,
                remaining: [card1.value, card2.value, card3.value],
            }),
            _ => HandRank::HighCard(HighCard {
                remaining: [
                    card1.value,
                    card2.value,
                    card3.value,
                    card4.value,
                    card5.value,
                ],
            }),
        }
    }
}

fn rank_hand(hand: &Hand) -> HandRank {
    let opt_straight = find_straight(hand);
    let opt_flush = find_flush(hand);
    match opt_flush {
        Some(flush) => match opt_straight {
            Some(straight) => match straight.highest {
                ACE => HandRank::RoyalFlush,
                v => HandRank::StraightFlush(StraightFlush { highest: v }),
            },
            None => HandRank::Flush(flush),
        },
        None => match opt_straight {
            Some(straight) => HandRank::Straight(straight),
            None => match find_four(hand) {
                Some(four) => HandRank::Four(four),
                None => rank_low_hand(hand),
            },
        },
    }
}

fn eval_line(line: &str) -> (HandRank, HandRank) {
    let (hx, hy) = parse_hands(&line.chars().collect_vec());
    (rank_hand(&hx), rank_hand(&hy))
}

pub fn euler054() -> String {
    crate::utils::input::get_input(54)
        .lines()
        .map(eval_line)
        .filter(|(ra, rb)| ra > rb)
        .count()
        .to_string()
}
